File: fastapi_app/routers/engagement.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Query

from fastapi_app.schemas.platform import EngagementRequest
from fastapi_app.security import require_api_key
from fastapi_app.services import engagement_service

router = APIRouter(prefix="/engagement", tags=["Engagement"])


@router.post("/{learner_id}")
def record_engagement(learner_id: str, payload: EngagementRequest, _: None = Depends(require_api_key)):
    return engagement_service.record_engagement(learner_id, payload.event_type, payload.metadata)


@router.get("/{learner_id}/heatmap")
def heatmap(
    learner_id: str,
    period: str = Query(default="all", pattern="^(7d|30d|all)$"),
    _: None = Depends(require_api_key),
):
    return engagement_service.get_heatmap(learner_id, period)


@router.get("/{learner_id}")
def engagement_metrics(
    learner_id: str,
    period: str = Query(default="7d", pattern="^(7d|30d|all)$"),
    _: None = Depends(require_api_key),
):
    return engagement_service.get_daily_metrics(learner_id, period)

File: frontend/src/api/engagement.ts
import { apiClient } from './client';

export async function recordEngagement(
  learnerId: string,
  eventType: string,
  metadata: Record<string, unknown> = {},
) {
  const { data } = await apiClient.post(`/engagement/${learnerId}`, {
    event_type: eventType,
    metadata,
  });
  return data;
}

export async function getHeatmap(learnerId: string, period: '7d' | '30d' | 'all' = 'all') {
  const { data } = await apiClient.get<Array<{ date: string; count: number }>>(
    `/engagement/${learnerId}/heatmap`,
    { params: { period } },
  );
  return data;
}

export async function getEngagementMetrics(learnerId: string, period: '7d' | '30d' | 'all' = '7d') {
  const { data } = await apiClient.get<
    Array<{ date: string; study_time: number; questions_answered: number }>
  >(`/engagement/${learnerId}`, { params: { period } });
  return data;
}
